### ISSUE TITLE
Package openevse setup failed.

### ISSUE THREAD

**Issue by mattbruman** [TonyApuzzo/home-assistant-config#6]
Followed everything but I am getting this error when trying to reboot HA
rest_cmd folder is in custom_components

Package openevse setup failed. Component rest_cmd Integration 'rest_cmd' not found.
Invalid config for [input_select]: [values] is an invalid option for [input_select]. Check: input_select->input_select->openevse_service_level->values. (See ?, line ?).

**Comment by mattbruman** [comment]
I hope to resolve this as I would like to be able to remotely disable the charging and start it again from HA. I want the ability to say to google Im leaving in a hour to start charging the car otherwise stick to charging in off hours.

**Comment by TonyApuzzo** [comment]
I at one point I was using a custom component for rest_cmd, I will review my setup and see if I can give you more advice on getting it working, it may be that my local install and what's in github have diverged too much -- I may have some time to look at it this evening.

**Comment by mattbruman** [comment]
> I at one point I was using a custom component for rest_cmd, I will review my setup and see if I can give you more advice on getting it working, it may be that my local install and what's in github have diverged too much -- I may have some time to look at it this evening.

I used your rest_cmd folder you have on here, was I supposed to use a different rest_cmd?

**Comment by TonyApuzzo** [comment]
You're correct -- You do need the rest_cmd custom component that is in the custom_components folder.

I just looked at my configuration and it is sync'd to github, so the master branch here should be working. I don't get any errors or warnings in my server log from openevse or the rest_cmd custom component.

I am not sure what is going wrong.

Are you running the latest version of Home Assistant?  The most recent check-in of this configuration requires 2021.9.4 

Do you see these warnings logged in the server log on startup?
```text
2021-09-08 11:02:58 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration rest_cmd which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2021-09-08 11:02:58 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration input_select which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2021-09-08 11:02:58 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration input_number which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
```

if not then maybe the custom component folder is in the wrong place or there's a syntax error or other problem in the files.

EDIT: my package requires the OpenEVSE v2.9.1 firmware on the old / original WiFi module that uses Huzzah ESP8266. Also it appears that there has been a fix for the official integration, I'm not 100% sure which WiFi modules are supported by that integration, but you could try it if you continue to struggle with this package.

**Comment by mattbruman** [comment]
Yup running latest version I copied those custom components into my custom_components folder
![image](https://user-images.githubusercontent.com/43623584/132609989-28ca4e2d-ac36-46fe-90d3-188c120b07fa.png)
I then copied your secrets example into my secrets and then edited it.
![image](https://user-images.githubusercontent.com/43623584/132610326-1ab1c68a-f86c-45ab-bfa4-dfc7ca151846.png)
and then added the openevse.yaml file to my packages, I already use packages so it was enabled.
![image](https://user-images.githubusercontent.com/43623584/132610543-58367a3f-3ee8-4b78-af03-c89a73d7bc03.png)
and then I check the config and I get an error so I cant even reboot to check if the customs worked
![image](https://user-images.githubusercontent.com/43623584/132610655-576d5ba3-4c1b-4779-a0a9-3d158d288885.png)

**Comment by mattbruman** [comment]
As for the fix yeah its kinda working but it creates a ton of errors in the log and I cant enable and disable charging from HA, I would rather a fully working mqtt option with feedback

**Comment by mattbruman** [comment]
Looks like I'm not the only one trying to get this to work atm

https://community.openenergymonitor.org/t/openevse-home-assistant-integration/7059/16

**Comment by TonyApuzzo** [comment]
Thank you for posting the troubleshooting.  Sorry I don't have any advice yet.

I'm going to create a new install on a spare Raspberry Pi and see if I can get it working (or failing) with a simplified configuration.

**Comment by TonyApuzzo** [comment]
OK, I built a new install of HA on my spare RPI 4, and I got the same error as you got when initially trying to Check Configuration or even restart HASS Core.

To debug this, I installed the add-on [Check Home Assistant configuration](https://github.com/home-assistant/addons/tree/master/check_config) and used that to verify the configuration.

That succeeded!

It turns out (I don't know WHY this is) that you have to reboot the host in order for the custom components to be initially detected: Supervisor -> System -> Host -> Reboot Host -> Reboot Host

I got this idea from this forum post: https://community.home-assistant.io/t/custom-components-doesnt-load-in-hassos/197908/8?u=a3a

Note that in order for the MQTT sensors to work you have to:
1. Have the MQTT integration configured
2. Be using an MQTT account that has permission to read from and write to the `openevse/#` topics (or whatever base topic you configure in the secrets.)
3. If you're using Home Assistant's add-on for MQTT broker, you have to restart the server after changing the ACL.

**Comment by mattbruman** [comment]
So I re added them back again and they are detected but they error out on boot

> 2021-09-10`` 11:07:12 ERROR (MainThread) [homeassistant.config] Invalid config for [input_select]: required key not provided @ data['input_select']. Got None. (See /config/configuration.yaml, line 2). Please check the docs at https://www.home-assistant.io/components/input_select

> 2021-09-10 11:07:12 ERROR (MainThread) [homeassistant.setup] Setup failed for input_select: Invalid config.

> 2021-09-10 11:07:12 ERROR (MainThread) [homeassistant.config] Invalid config for [input_number]: required key not provided @ data['input_number']. Got None. (See /config/configuration.yaml, line 2). Please check the docs at https://www.home-assistant.io/components/input_number

> 2021-09-10 11:07:12 ERROR (MainThread) [homeassistant.setup] Setup failed for input_number: Invalid config.

> 2021-09-10 11:07:20 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of default_config. Setup failed for dependencies: input_number, input_select

> 2021-09-10 11:07:20 ERROR (MainThread) [homeassistant.setup] Setup failed for default_config: Could not set up all dependencies.

**Comment by TonyApuzzo** [comment]
I've created a super-simple package that ONLY has MQTT and doesn't need any custom components. It doesn't use the REST API at all which is the source of all the complexity in my previous package.

This package supports only Enable/Disable, Status and Current Charging Current, you can use the `input_boolean.openevse_enabled` entity to enable/disable charging on a schedule or via any automation that you want. You can see it in action.  Note that this does require that you use the "openevse" topic in your WiFi configuration for the OpenEVSE MQTT panel.

You can delete all the stuff you got from me before, including the secrets and just add the below to a new file 

**/config/packages/openevse2.yaml**

```yaml
#
# Simple OpenEVSE package
# Supports ONLY MQTT interactions
#

input_boolean:
  openevse_enabled:
    name: Enable / Disable OpenEVSE
    icon: mdi:power-cycle

sensor:
  - platform: mqtt
    name: OpenEVSE MQTT State
    icon: mdi:car-electric
    state_topic: >-
      openevse/state
  - platform: mqtt
    name: OpenEVSE Current Now
    icon: mdi:current-ac
    state_topic: >-
      openevse/amp
    device_class: current
    state_class: measurement
    unit_of_measurement: "A"
    value_template: >-
      {{- (value|float / 1000.0) | round(1) -}}
template:
  - trigger:
    - platform: state
      entity_id: sensor.openevse_mqtt_state
    sensor:
      - name: OpenEVSE State
        state: >-
          {%- set value = states(trigger.entity_id) -%}
          {%- if value | regex_match('^[01]$') -%}
            Not Connected
          {%- elif value | regex_match('^2$') -%}
            Connected
          {%- elif value | regex_match('^3$') -%}
            Charging
          {%- elif value | regex_match('^([456789]|10)$') -%}
            Error ({{- value -}})
          {%- elif value | regex_match('^254$') -%}
            Sleeping
          {%- elif value | regex_match('^255$') -%}
            Disabled
          {%- else -%}
            Unknown ({{- value -}})
          {%- endif -%}
        icon: >-
          {%- set value = states(trigger.entity_id) -%}
          {%- if value | regex_match('^[01]$') -%}
            mdi:power-plug-off
          {%- elif value | regex_match('^2$') -%}
            mdi:car-electric
          {%- elif value | regex_match('^3$') -%}
            mdi:battery-charging
          {%- elif value | regex_match('^([456789]|10)$') -%}
            mdi:battery-alert
          {%- elif value | regex_match('^254$') -%}
            mdi:sleep
          {%- elif value | regex_match('^255$') -%}
            mdi:power-off
          {%- else -%}
            mdi:battery-unknown
          {%- endif -%}

automation openevse2:
- alias: "Detect OpenEVSE Enablement"
  trigger:
    platform: mqtt
    topic: 'openevse/state'
  action:
    service: >
      input_boolean.turn_{{- 'off' if trigger.payload|int == 255 or trigger.payload|int == 254 else 'on' -}}
    target:
      entity_id: input_boolean.openevse_enabled

- alias: "Set OpenEVSE Enablement"
  trigger:
    - platform: state
      entity_id: input_boolean.openevse_enabled
      from: 'on'
      to: 'off'
    - platform: state
      entity_id: input_boolean.openevse_enabled
      from: 'off'
      to: 'on'
  action:
    service: mqtt.publish
    data:
      topic: >
        openevse/rapi/in/{{- '$FE' if is_state('input_boolean.openevse_enabled', 'on') else '$FD' -}}
```

![image](https://user-images.githubusercontent.com/341037/132924652-39af727c-9fbf-4e74-9859-135d1cc117dd.png)

**Comment by TonyApuzzo** [comment]
I've decided that this MQTT only package is a much cleaner way to set things up and I'm going to switch to implementing this in my main configuration and will eventually publish this as a blueprint.

You can get the latest via a gist, I will update this as I go.

https://gist.github.com/TonyApuzzo/c0fe22458d4dae0effb47b951b464484#file-openevse2-yaml

NOTE: 
The Session Energy and Service Level sensors aren't updated automatically, you have to manually publish the corresponding RAPI command $GU and $GE respectively. As I figure out how to do that I'll update the package.

**Comment by mattbruman** [comment]
Thanks for fixing things, I'm gonna give it a go at installing today. Ill post a update.

**Comment by mattbruman** [comment]
Nice with both official integration and yours running I get almost all the functionality I desire. Only thing that would be nice is the Energy Limit setting so I can do a automation like "Add 10 Miles" which could cut off after 4kWh because I can translate kWh to miles easily. Only reason I want that is I have cheaper energy at night and if I just happen to need extra range in the day I could just add as much as I need at the more expensive rate. But right now I just have a automation to turn off and on the charger after a set time which is essentially the same thing just not as accurate. 

I drive the Pacifica hybrid which only has around 40 mile range on a good day so I have to juggle range much more than a full electric.
![image](https://user-images.githubusercontent.com/43623584/133128616-ebc1f73f-62f7-410d-af1a-872e1d880b77.png)

**Comment by TonyApuzzo** [comment]
I will look at adding energy limit configuration, shouldn't be too difficult since there is an MQTT api for it.

A few of the sensors are really only for 'internal' use or debugging, e.g.: RAPI Out, RAPI In, 

I added scripts for triggering the Session Energy data update and the Settings Flags (from which you can derive Service L2/L1/Auto) -- I intend to figure out some way of periodically triggering those service calls to keep these values somewhat up-to-date since the unit itself doesn't publish these.

**Comment by TonyApuzzo** [comment]
I updated the gist with energy limit `OpenEVSE Charge Limit` input.

The API for OpenEVSE is weird with the charge limit feature.  A charge limit can only be set to the EVSE when the plug is connected.

So what I did was allow you to set a charge limit whenever you want, but it will "fail" to set the charge limit to the device, however if the charger goes into Connected or Charging state, then it will push a charge-limit value to the EVSE.

Currently I do not reset the charge limit back to zero, so if you turn the charger back on, it will re-apply whatever charge limit was set previously.  If you want to allow a full charge, then set the charge limit to 0.

Oh yeah, my car is weird in that when charging stops (or the device goes into 'Sleeping' state), then re-enabling the charger doesn't start charging again.  I have to to an extra cycle, e.g., enable > disable > enable and then my car will start charging again. I think this is a quirk with my Kia Soul EV, but it could be the charger's fault, I won't know unless you see the same thing or not.
